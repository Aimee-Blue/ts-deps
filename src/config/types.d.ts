type Environments =
  | 'chrome'
  | 'opera'
  | 'edge'
  | 'firefox'
  | 'safari'
  | 'ie'
  | 'ios'
  | 'android'
  | 'node'
  | 'electron';

type Targets = {
  esmodules?: boolean;
  node?: string | 'current' | true;
  safari?: string | 'tp';
  browsers?: string | string[];
  [key in Environments]?: string;
};

interface BabelPresetEnvConfig {
  targets?: string | Array<string> | Targets;
  spec?: boolean;
  loose?: boolean;
  modules?: 'amd' | 'umd' | 'systemjs' | 'commonjs' | 'cjs' | 'auto' | false;
  debug?: boolean;
  include?: Array<string | RegExp>;
  exclude?: Array<string | RegExp>;
  useBuiltIns?: 'usage' | 'entry' | false;
  corejs?: 2 | 3 | { version: 2 | 3; proposals: boolean };
  forceAllTransformations?: boolean;
  configPath?: string;
  ignoreBrowserslistConfig?: boolean;
  shippedProposals?: boolean;
}

interface BabelConfigParams {
  aliases: Options['aliases'];
  nodeVersion: string;
  presetEnvConfig?: (
    defaultConfig: BabelPresetEnvConfig
  ) => BabelPresetEnvConfig;
}

interface JestConfigParams {
  aliases: Options['aliases'];
  nodeVersion: string;
}

interface TsConfigParams {
  aliases: Options['aliases'];
  baseConfigLocation: string;
}

interface EsLintConfigParams {
  env?: Record<string, boolean>;
  extends?: string[];
  ignorePatterns?: string[];
  parserOptions?: object;
  rules?: object;
  [key: string]: unknown;
}

interface EsLintConfig {
  root?: EsLintConfigParams;
  src?: EsLintConfigParams;
}
