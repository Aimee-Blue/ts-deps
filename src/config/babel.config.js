// @ts-check
'use strict'

const defaults = require('../defaults')
const { trimPathSeparator } = require('../helpers')
const { options } = require('../options')

/**
 * @param {Partial<IBabelConfigParams>} paramsRaw
 */
const babelConfig = (paramsRaw = {}) => {
  const opts = options()
  const { aliases, nodeVersion } = {
    aliases: opts.aliases,
    nodeVersion: opts.nodeVersion || defaults.nodeVersion,
    ...paramsRaw,
  }
  const src = defaults.rootDir

  const config = {
    presets: [
      [
        '@babel/preset-env',
        {
          targets: {
            node: nodeVersion,
          },
        },
      ],
      '@babel/preset-typescript',
    ],
    plugins: [
      !!aliases && [
        'module-resolver',
        {
          root: [src],
          alias: Object.keys(aliases).reduce((acc, key) => {
            return {
              ...acc,
              [key]: trimPathSeparator(aliases[key]),
            }
          }, {}),
        },
      ],
      '@babel/proposal-class-properties',
    ].filter(Boolean),
    ignore: [`${src}/**/*.d.ts`, `${src}/**/*.json`],
  }

  return config
}

module.exports = {
  __esModule: {
    value: true,
  },
  default: babelConfig(),
  babelConfig,
}
