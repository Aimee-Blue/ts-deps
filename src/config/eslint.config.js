// @ts-check
import restrictedGlobals from 'confusing-browser-globals';

/**
 * @param {EsLintConfig} params
 */
export function eslintConfig(params) {
  const rootJavaScriptConfig = {
    root: true,

    env: {
      browser: false,
      es6: true,
      node: true,
      ...params?.root?.env,
    },

    parser: 'babel-eslint',

    plugins: ['import'],

    extends: [
      'eslint:recommended',
      'prettier',
      ...(params?.root?.extends || []),
    ],

    parserOptions: {
      ecmaVersion: 2018,
      ...params?.root?.parserOptions,
    },

    rules: {
      'array-callback-return': 'warn',
      'default-case': ['warn', { commentPattern: '^no default$' }],
      'dot-location': ['warn', 'property'],
      eqeqeq: ['warn', 'smart'],
      'new-parens': 'warn',
      'no-array-constructor': 'warn',
      'no-caller': 'warn',
      'no-cond-assign': ['warn', 'except-parens'],
      'no-const-assign': 'warn',
      'no-control-regex': 'warn',
      'no-delete-var': 'warn',
      'no-dupe-args': 'warn',
      'no-dupe-class-members': 'warn',
      'no-dupe-keys': 'warn',
      'no-duplicate-case': 'warn',
      'no-empty-character-class': 'warn',
      'no-empty-pattern': 'warn',
      'no-eval': 'warn',
      'no-ex-assign': 'warn',
      'no-extend-native': 'warn',
      'no-extra-bind': 'warn',
      'no-extra-label': 'warn',
      'no-fallthrough': 'warn',
      'no-func-assign': 'warn',
      'no-implied-eval': 'warn',
      'no-invalid-regexp': 'warn',
      'no-iterator': 'warn',
      'no-label-var': 'warn',
      'no-labels': ['warn', { allowLoop: true, allowSwitch: false }],
      'no-lone-blocks': 'warn',
      'no-loop-func': 'warn',
      'no-mixed-operators': [
        'warn',
        {
          groups: [
            ['&', '|', '^', '~', '<<', '>>', '>>>'],
            ['==', '!=', '===', '!==', '>', '>=', '<', '<='],
            ['&&', '||'],
            ['in', 'instanceof'],
          ],
          allowSamePrecedence: false,
        },
      ],
      'no-multi-str': 'warn',
      'no-native-reassign': 'warn',
      'no-negated-in-lhs': 'warn',
      'no-new-func': 'warn',
      'no-new-object': 'warn',
      'no-new-symbol': 'warn',
      'no-new-wrappers': 'warn',
      'no-obj-calls': 'warn',
      'no-octal': 'warn',
      'no-octal-escape': 'warn',
      // TODO: Remove this option in the next major release of CRA.
      // https://eslint.org/docs/user-guide/migrating-to-6.0.0#-the-no-redeclare-rule-is-now-more-strict-by-default
      'no-redeclare': ['warn', { builtinGlobals: false }],
      'no-regex-spaces': 'warn',
      'no-restricted-syntax': ['warn', 'WithStatement'],
      'no-script-url': 'warn',
      'no-self-assign': 'warn',
      'no-self-compare': 'warn',
      'no-sequences': 'warn',
      'no-shadow-restricted-names': 'warn',
      'no-sparse-arrays': 'warn',
      'no-template-curly-in-string': 'warn',
      'no-this-before-super': 'warn',
      'no-throw-literal': 'warn',
      'no-undef': 'error',
      'no-restricted-globals': ['error'].concat(restrictedGlobals),
      'no-unreachable': 'warn',
      'no-unused-expressions': [
        'error',
        {
          allowShortCircuit: true,
          allowTernary: true,
          allowTaggedTemplates: true,
        },
      ],
      'no-unused-labels': 'warn',
      'no-use-before-define': [
        'warn',
        {
          functions: false,
          classes: false,
          variables: false,
        },
      ],
      'no-useless-computed-key': 'warn',
      'no-useless-concat': 'warn',
      'no-useless-constructor': 'warn',
      'no-useless-escape': 'warn',
      'no-useless-rename': [
        'warn',
        {
          ignoreDestructuring: false,
          ignoreImport: false,
          ignoreExport: false,
        },
      ],
      'no-with': 'warn',
      'no-whitespace-before-property': 'warn',
      'require-yield': 'warn',
      'rest-spread-spacing': ['warn', 'never'],
      'unicode-bom': ['warn', 'never'],
      'use-isnan': 'warn',
      'valid-typeof': 'warn',
      'no-restricted-properties': [
        'error',
        {
          object: 'require',
          property: 'ensure',
          message:
            'Please use import() instead. More info: https://facebook.github.io/create-react-app/docs/code-splitting',
        },
        {
          object: 'System',
          property: 'import',
          message:
            'Please use import() instead. More info: https://facebook.github.io/create-react-app/docs/code-splitting',
        },
      ],
      'getter-return': 'warn',

      'import/first': 'error',
      'import/no-amd': 'error',
      'import/no-webpack-loader-syntax': 'error',

      'no-unused-vars': [
        'error',
        {
          ignoreRestSiblings: true,
          argsIgnorePattern: '_.*',
        },
      ],
      ...params?.root?.rules,
    },
  };

  const typeScript = {
    parser: '@typescript-eslint/parser',
    env: {
      browser: false,
      es6: true,
      node: true,
      jest: true,
      ...params?.src?.env,
    },
    parserOptions: {
      project: './tsconfig.json',
      ecmaVersion: 2018,
      sourceType: 'module',
      ecmaFeatures: {
        jsx: true,
      },
      warnOnUnsupportedTypeScriptVersion: true,
      ...params?.src?.parserOptions,
    },
    plugins: ['import', '@typescript-eslint'],
    extends: [
      'eslint:recommended',
      'plugin:@typescript-eslint/eslint-recommended',
      'plugin:@typescript-eslint/recommended',
      'plugin:@typescript-eslint/recommended-requiring-type-checking',
      'prettier',
      'prettier/@typescript-eslint',
      ...(params?.src?.extends ?? []),
    ],
    rules: {
      strict: ['warn', 'never'],
      // TypeScript's `noFallthroughCasesInSwitch` option is more robust (#6906)
      'default-case': 'off',
      // 'tsc' already handles this (https://github.com/typescript-eslint/typescript-eslint/issues/291)
      'no-dupe-class-members': 'off',
      // 'tsc' already handles this (https://github.com/typescript-eslint/typescript-eslint/issues/477)
      'no-undef': 'off',
      // Add TypeScript specific rules (and turn off ESLint equivalents)
      '@typescript-eslint/consistent-type-assertions': 'warn',
      'no-array-constructor': 'off',
      '@typescript-eslint/no-array-constructor': 'warn',
      '@typescript-eslint/no-namespace': 'error',
      'no-use-before-define': 'off',
      '@typescript-eslint/no-use-before-define': [
        'warn',
        {
          functions: false,
          classes: false,
          variables: false,
          typedefs: false,
        },
      ],
      'no-unused-expressions': 'off',
      '@typescript-eslint/no-unused-expressions': [
        'error',
        {
          allowShortCircuit: true,
          allowTernary: true,
          allowTaggedTemplates: true,
        },
      ],
      'no-unused-vars': 'off',
      '@typescript-eslint/no-unused-vars': [
        'error',
        {
          ignoreRestSiblings: true,
          argsIgnorePattern: '_.*',
        },
      ],
      'no-useless-constructor': 'off',
      '@typescript-eslint/no-useless-constructor': 'warn',
      '@typescript-eslint/explicit-function-return-type': 'off',
      '@typescript-eslint/interface-name-prefix': 'warn',
      '@typescript-eslint/require-await': 'warn',
      '@typescript-eslint/prefer-regexp-exec': 'warn',

      ...params?.src?.rules,
    },

    overrides: [
      {
        files: ['src/**/__tests__/**/*.ts?(x)', 'src/**/__tests__/**/*.js?(x)'],
        rules: {
          '@typescript-eslint/ban-ts-ignore': 'off',
          '@typescript-eslint/no-explicit-any': 'off',
          '@typescript-eslint/require-await': 'off',
          ...params?.tests?.rules,
        },
      },
    ],
  };

  return {
    ...rootJavaScriptConfig,
    overrides: [
      {
        files: ['src/**/*.ts?(x)', 'src/**/*.js?(x)'],
        ...typeScript,
      },
    ],
  };
}
