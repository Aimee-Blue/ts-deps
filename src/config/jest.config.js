// @ts-check
'use strict'

const defaults = require('../defaults')
const { options } = require('../options')

const {
  ensureStartsWithPathSeparator,
  trimPathSeparator,
} = require('../helpers')

const jestConfig = (
  { aliases, isIntegrationTest = false } = { ...options() }
) => {
  const src = ensureStartsWithPathSeparator(defaults.rootDir)
  const lib = trimPathSeparator(defaults.outDir)
  const extensions = defaults.extensions
  const exts = defaults.extensions.join(',')
  const pipedExts = defaults.extensions.join('|')
  const moduleNameMapper =
    aliases &&
    Object.keys(aliases).reduce((acc, key) => {
      const alias = trimPathSeparator(aliases[key])
      return {
        ...acc,
        [`^${key}$`]: `<rootDir>/${alias}`,
        [`^${key}/(.*)$`]: `<rootDir>/${alias}/$1`,
      }
    }, {})

  const integrationTestMatch = `<rootDir>/${
    defaults.integrationTestsGlob
  }/*.test.{${exts}}`
  const unitTestMatch = `<rootDir>/${defaults.unitTestsGlob}/*.test.{${exts}}`

  return {
    testEnvironment: 'node',
    testMatch: isIntegrationTest ? [integrationTestMatch] : [unitTestMatch],
    cacheDirectory: '.jest-cache',
    collectCoverageFrom: [
      `<rootDir>${src}/**/*.{${exts}}`,
      '!**/node_modules/**',
      '!**/*.d.ts',
    ],
    moduleFileExtensions: extensions,
    setupFiles: [],
    ...(moduleNameMapper && { moduleNameMapper }),
    testPathIgnorePatterns: ['/node_modules/', lib],
    transform: {
      [`^.+\\.(${pipedExts})$`]: '<rootDir>/node_modules/babel-jest',
    },
    modulePathIgnorePatterns: [lib],
    globals: {
      INTEGRATION_TEST: false,
    },
  }
}

module.exports = {
  jestConfig,
}
