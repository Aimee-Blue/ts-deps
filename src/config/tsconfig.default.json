{
  "compilerOptions": {
    "lib": ["es2017"],
    "target": "es2017",
    "baseUrl": ".",
    "module": "commonjs",
    "sourceMap": false,
    "moduleResolution": "node",
    "allowJs": true,
    "checkJs": false,
    "skipLibCheck": true,
    "strict": true,
    "noImplicitAny": true,
    "suppressImplicitAnyIndexErrors": false,
    "strictNullChecks": true,
    "strictFunctionTypes": false,
    "strictPropertyInitialization": true,
    "noImplicitThis": true,
    "alwaysStrict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "forceConsistentCasingInFileNames": true,
    "isolatedModules": false,
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "noEmit": true,
    "jsx": "preserve"
  },
  "compileOnSave": false
}
