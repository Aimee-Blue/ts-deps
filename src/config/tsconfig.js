// @ts-check
const { options } = require('../options');
const { trimPathSeparator } = require('../helpers');
const defaults = require('../defaults');

/**
 * @param {Partial<ITsConfigParams>} paramsRaw
 */
const tsConfig = (paramsRaw = {}) => {
  const opts = options();
  const { aliases, baseConfigLocation } = {
    aliases: opts.aliases,
    baseConfigLocation:
      opts.baseTsConfigLocation ||
      './node_modules/ts-deps/lib/config/tsconfig.default.json',
    ...paramsRaw,
  };
  const src = defaults.rootDir;
  const lib = defaults.outDir;
  const paths =
    aliases &&
    Object.keys(aliases).reduce((acc, key) => {
      const alias = trimPathSeparator(aliases[key]);
      return {
        ...acc,
        [key]: [alias],
        [`${key}/*`]: [`${alias}/*`],
      };
    }, {});

  return {
    extends: baseConfigLocation,
    compilerOptions: {
      baseUrl: '.',
      paths,
      outDir: lib,
    },
    include: [src],
  };
};

/**
 * @param {Partial<ITsConfigParams>} paramsRaw
 */
const tsConfigDeclarations = (paramsRaw = {}) => {
  /**
   * @type {ReturnType<typeof tsConfig> & { exclude?: string[] }}
   */
  const defConfig = tsConfig(paramsRaw);

  return {
    ...defConfig,
    compilerOptions: {
      ...defConfig.compilerOptions,
      allowJs: false,
      noEmit: false,
      declaration: true,
      emitDeclarationOnly: true,
    },
    exclude: [
      ...((Array.isArray(defConfig.exclude) && defConfig.exclude) || []),
      defaults.unitTestsGlob,
      defaults.integrationTestsGlob,
    ],
  };
};

module.exports = {
  tsConfig,
  tsConfigDeclarations,
};
