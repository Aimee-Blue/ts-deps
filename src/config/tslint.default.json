{
  "extends": ["tslint:recommended", "tslint-config-prettier"],
  "jsRules": {
    "object-literal-sort-keys": false,
    "ordered-imports": false,
    "variable-name": [true, "allow-leading-underscore", "allow-pascal-case"],
    "no-console": false
  },
  "rules": {
    "object-literal-sort-keys": false,
    "ordered-imports": false,
    "variable-name": [true, "allow-leading-underscore", "allow-pascal-case"],
    "no-console": false,
    "strict-boolean-expressions": [
      true,
      "allow-string",
      "allow-boolean-or-undefined",
      "allow-undefined-union",
      "allow-null-union",
      "allow-mix",
      "ignore-rhs"
    ],
    "no-unsafe-any": true,
    "no-any": true,
    "await-promise": true
  },
  "linterOptions": {
    "exclude": ["src/**/*.test.*"]
  }
}
