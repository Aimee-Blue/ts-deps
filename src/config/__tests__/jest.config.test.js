import { jestConfig } from '../jest.config';

describe('given no options', () => {
  it('should work', () => {
    const defaultConfig = jestConfig();
    expect(defaultConfig).toMatchSnapshot();
  });
});

describe('given aliases', () => {
  it('should work', () => {
    const defaultConfig = jestConfig({
      aliases: {
        ['@shared']: '/src/shared',
      },
    });
    expect(defaultConfig).toMatchSnapshot();
  });
});

describe('given integration tests', () => {
  it('should work', () => {
    const defaultConfig = jestConfig({
      isIntegrationTest: true,
    });
    expect(defaultConfig).toMatchSnapshot();
  });
});
