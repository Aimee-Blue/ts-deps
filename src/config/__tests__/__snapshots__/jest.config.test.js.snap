// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given aliases should work 1`] = `
Object {
  "cacheDirectory": ".jest-cache",
  "collectCoverageFrom": Array [
    "<rootDir>/src/**/*.{js,jsx,ts,tsx,json}",
    "!**/node_modules/**",
    "!src/__integration-tests__/**",
    "!src/**/__tests__/**",
    "!<rootDir>/src/**/*.d.ts",
    "!<rootDir>/src/**/*.json",
  ],
  "coverageDirectory": "coverage-unit",
  "coverageReporters": Array [
    "json",
    "lcov",
    "text",
  ],
  "globals": Object {
    "INTEGRATION_TEST": false,
  },
  "moduleFileExtensions": Array [
    "js",
    "jsx",
    "ts",
    "tsx",
    "json",
  ],
  "moduleNameMapper": Object {
    "^@shared$": "<rootDir>/src/shared",
    "^@shared/(.*)$": "<rootDir>/src/shared/$1",
  },
  "modulePathIgnorePatterns": Array [
    "/lib/",
  ],
  "setupFiles": Array [],
  "testEnvironment": "node",
  "testMatch": Array [
    "<rootDir>/src/**/__tests__/**/*.test.{js,jsx,ts,tsx,json}",
  ],
  "testPathIgnorePatterns": Array [
    "/node_modules/",
    "/lib/",
  ],
  "transformIgnorePatterns": Array [
    "node_modules",
    "/lib/",
    ".*\\\\.json",
    ".*\\\\.d\\\\.ts",
  ],
}
`;

exports[`given integration tests should work 1`] = `
Object {
  "cacheDirectory": ".jest-cache",
  "collectCoverageFrom": Array [
    "<rootDir>/src/**/*.{js,jsx,ts,tsx,json}",
    "!**/node_modules/**",
    "!src/__integration-tests__/**",
    "!src/**/__tests__/**",
    "!<rootDir>/src/**/*.d.ts",
    "!<rootDir>/src/**/*.json",
  ],
  "coverageDirectory": "coverage-integration",
  "coverageReporters": Array [
    "json",
    "lcov",
    "text",
  ],
  "globals": Object {
    "INTEGRATION_TEST": false,
  },
  "moduleFileExtensions": Array [
    "js",
    "jsx",
    "ts",
    "tsx",
    "json",
  ],
  "moduleNameMapper": Object {
    "^ts-deps$": "<rootDir>/./src",
    "^ts-deps/(.*)$": "<rootDir>/./src/$1",
  },
  "modulePathIgnorePatterns": Array [
    "/lib/",
  ],
  "setupFiles": Array [],
  "testEnvironment": "node",
  "testMatch": Array [
    "<rootDir>/src/__integration-tests__/**/*.test.{js,jsx,ts,tsx,json}",
  ],
  "testPathIgnorePatterns": Array [
    "/node_modules/",
    "/lib/",
  ],
  "transformIgnorePatterns": Array [
    "node_modules",
    "/lib/",
    ".*\\\\.json",
    ".*\\\\.d\\\\.ts",
  ],
}
`;

exports[`given no options should work 1`] = `
Object {
  "cacheDirectory": ".jest-cache",
  "collectCoverageFrom": Array [
    "<rootDir>/src/**/*.{js,jsx,ts,tsx,json}",
    "!**/node_modules/**",
    "!src/__integration-tests__/**",
    "!src/**/__tests__/**",
    "!<rootDir>/src/**/*.d.ts",
    "!<rootDir>/src/**/*.json",
  ],
  "coverageDirectory": "coverage-unit",
  "coverageReporters": Array [
    "json",
    "lcov",
    "text",
  ],
  "globals": Object {
    "INTEGRATION_TEST": false,
  },
  "moduleFileExtensions": Array [
    "js",
    "jsx",
    "ts",
    "tsx",
    "json",
  ],
  "moduleNameMapper": Object {
    "^ts-deps$": "<rootDir>/./src",
    "^ts-deps/(.*)$": "<rootDir>/./src/$1",
  },
  "modulePathIgnorePatterns": Array [
    "/lib/",
  ],
  "setupFiles": Array [],
  "testEnvironment": "node",
  "testMatch": Array [
    "<rootDir>/src/**/__tests__/**/*.test.{js,jsx,ts,tsx,json}",
  ],
  "testPathIgnorePatterns": Array [
    "/node_modules/",
    "/lib/",
  ],
  "transformIgnorePatterns": Array [
    "node_modules",
    "/lib/",
    ".*\\\\.json",
    ".*\\\\.d\\\\.ts",
  ],
}
`;
